jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleHelper.java

```diff
@@ -0,0 +1,699 @@
+				break;
+			case 2: /* getStatic */
+				result = lookup.findStaticGetter(referenceClazz, name, resolveFieldHandleHelper(typeDescriptor, lookup, loader));
+				break;
+			case 3: /* putField */
+				result = lookup.findSetter(referenceClazz, name, resolveFieldHandleHelper(typeDescriptor, lookup, loader));
+				break;
+			case 4: /* putStatic */
+				result = lookup.findStaticSetter(referenceClazz, name, resolveFieldHandleHelper(typeDescriptor, lookup, loader));
+				break;
+			case 5: /* invokeVirtual */
+				type = MethodType.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+				lookup.accessCheckArgRetTypes(type);
+				result = lookup.findVirtual(referenceClazz, name, type);
+				break;
+			case 6: /* invokeStatic */
+				type = MethodType.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+				lookup.accessCheckArgRetTypes(type);
+				result = lookup.findStatic(referenceClazz, name, type);
+				break;
+			case 7: /* invokeSpecial */ 
+				type = MethodType.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+				lookup.accessCheckArgRetTypes(type);
+				result = lookup.findSpecial(referenceClazz, name, type, currentClass);
+				break;
+			case 8: /* newInvokeSpecial */
+				type = MethodType.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+				lookup.accessCheckArgRetTypes(type);
+				result = lookup.findConstructor(referenceClazz, type);
+				break;
+			case 9: /* invokeInterface */
+				type = MethodType.vmResolveFromMethodDescriptorString(typeDescriptor, loader, null);
+				lookup.accessCheckArgRetTypes(type);
+				result = lookup.findVirtual(referenceClazz, name, type);
+				break;
+			default:
+				/* Can never happen */
+				throw new UnsupportedOperationException();
+			}
+			return result;
+		} catch (IllegalAccessException iae) {
+			// Java spec expects an IllegalAccessError instead of IllegalAccessException thrown when an application attempts 
+			// (not reflectively) to access or modify a field, or to invoke a method that it doesn't have access to.
+			throw new IllegalAccessError(iae.getMessage()).initCause(iae);
+		}
+/*[ENDIF] OPENJDK_METHODHANDLES*/
+	}
+	
+	/* Convert the field type descriptor into a MethodType so we can reuse the parsing logic in 
+	 * #fromMethodDescriptorString().  The verifier checks to ensure that the typeDescriptor is
+	 * a valid field descriptor so adding the "()V" around it is valid.
+	 */
+	private static final Class<?> resolveFieldHandleHelper(String typeDescriptor, Lookup lookup, ClassLoader loader) throws Throwable {
+/*[IF OPENJDK_METHODHANDLES]*/
+		MethodType mt = MethodType.vmResolveFromMethodDescriptorString("(" + typeDescriptor + ")V", loader, null); //$NON-NLS-1$ //$NON-NLS-2$
+		return mt.parameterType(0);
+/*[ELSE]*/
+		MethodType mt = MethodType.vmResolveFromMethodDescriptorString("(" + typeDescriptor + ")V", loader, null); //$NON-NLS-1$ //$NON-NLS-2$
+		lookup.accessCheckArgRetTypes(mt);
```

Comment:
Why does OpenJDK methodHandles not need this access check?

Reply:
And similar to the comment about - make it clear what the difference is:

```
/*[IF OPENJDK_METHODHANDLES]*/
/* TODO: Investigate if an equivalent for lookup.accessCheckArgRetTypes is
 * needed for OJDK MH. If needed, find an alternative for
 * lookup.accessCheckArgRetTypes.
 */
/*[ELSE]*/
lookup.accessCheckArgRetTypes(type);
/*[ENDIF]*/
```
and keep the rest of the code common

Reply:
Not known yet. TBD.

Reply:
Comment added. `sendResolveMethodHandle` and `resolveFieldHandleHelper` shared between J9 and OJDK MH using preprocessor flags.